onnxruntime/core/session/inference_session.cc

```diff
@@ -90,42 +90,49 @@ inline std::basic_string<T> GetCurrentTimeString() {
   OrtStrftime<T>(time_str, sizeof(time_str), GetDateFormatString<T>(), &local_tm);
   return std::basic_string<T>(time_str);
 }
+
+concurrency::ThreadPool* CreateThreadPool(int size) {
+  if (size == 1)
+    return nullptr;
+  if (size <= 0)
+    size = std::max<int>(std::thread::hardware_concurrency() - 1, 1);
+  return new concurrency::ThreadPool("SESSION", size);
+}
+
 }  // namespace
 
 InferenceSession::InferenceSession(const SessionOptions& session_options,
-                                   logging::LoggingManager* logging_manager)
+                                   logging::LoggingManager* logging_manager,
+                                   concurrency::ThreadPool* thread_pool)
     : session_options_{session_options},
-      graph_transformation_mgr_{session_options_.max_num_graph_transformation_steps},
+      graph_transformation_mgr_{session_options.max_num_graph_transformation_steps},
       logging_manager_{logging_manager},
+      thread_pool_(thread_pool),
       session_state_(execution_providers_,
-                     session_options.enable_mem_pattern && session_options.enable_sequential_execution),
+                     session_options.enable_mem_pattern && session_options.enable_sequential_execution,
+                     thread_pool_),
       insert_cast_transformer_{"CastFloat16Transformer"} {
   ORT_ENFORCE(Environment::IsInitialized(),
               "Environment must be initialized before creating an InferenceSession.");
 
   InitLogger(logging_manager);
 
   session_state_.SetDataTransferMgr(&data_transfer_mgr_);
-
-  // The threadpool is currently evolving.  We will always create a per session threadpool.
-  // Beyond this, we will create a global thread pool to share across sessions.
-  {
-    int pool_size = session_options_.session_thread_pool_size <= 0
-                        ? std::thread::hardware_concurrency() / 2
-                        : session_options_.session_thread_pool_size;
-
-    thread_pool_ = std::make_unique<onnxruntime::concurrency::ThreadPool>("SESSION", pool_size);
-  }
-
-  session_state_.SetThreadPool(thread_pool_.get());
   session_profiler_.Initialize(session_logger_);
   session_state_.SetProfiler(session_profiler_);
   if (session_options.enable_profiling) {
     StartProfiling(session_options.profile_file_prefix);
   }
 }
 
+InferenceSession::InferenceSession(const SessionOptions& session_options,
+                                   logging::LoggingManager* logging_manager) : InferenceSession(session_options, logging_manager, CreateThreadPool(session_options.session_thread_pool_size)) {
+  //Even it's true, the threadpool could be a NULL pointer
+  own_threadpool_ = true;
```

Comment:
can this be part of the initialization list too?

Reply:
It's hard to make the order correct.